Common/Python/PythonInitializer.cs

```diff
@@ -57,31 +60,82 @@ public static void Initialize()
         /// <summary>
         /// Adds directories to the python path at runtime
         /// </summary>
-        public static void AddPythonPaths(IEnumerable<string> paths)
+        public static void AddPythonPaths(IEnumerable<string> paths, bool prepend = false)
         {
-            if (paths == null)
+            // Filter out any paths that are already on our Python path
+            paths = paths?.Where(x => !pathCache.Contains(x.Replace('\\', '/'))).ToList();
+            if (paths == null || !paths.Any())
             {
                 return;
             }
 
+            // Add these paths to our pending additions
+            if (prepend)
+            {
+                _pendingStartOfPathAdditions.AddRange(paths);
+            }
+            else
+            {
+                _pendingEndOfPathAdditions.AddRange(paths);
+            }
+
             if (_isInitialized)
             {
                 using (Py.GIL())
                 {
-                    _pendingPathAdditions.AddRange(paths);
+                    dynamic sys = Py.Import("sys");
+                    var locals = new PyDict();
+                    locals.SetItem("sys", sys);
+
+                    // Insert any pending start of path additions
+                    if (!_pendingStartOfPathAdditions.IsNullOrEmpty())
+                    {
+                        PythonEngine.Exec(string.Join(";", _pendingStartOfPathAdditions.Select(s => $"sys.path.insert(0, '{s}')"))
+                            .Replace('\\', '/'), null, locals.Handle);
+
+                        _pendingStartOfPathAdditions.Clear();
+                    }
 
-                    // Generate the python code to add these to our path and execute
-                    var code = string.Join(";", _pendingPathAdditions.Select(s => $"sys.path.append('{s}')"))
-                        .Replace('\\', '/');
+                    // Append any end of path additions
+                    if (!_pendingEndOfPathAdditions.IsNullOrEmpty())
+                    {
+                        PythonEngine.Exec(string.Join(";", _pendingEndOfPathAdditions.Select(s => $"sys.path.append('{s}')"))
+                            .Replace('\\', '/'), null, locals.Handle);
 
-                    PythonEngine.Exec($"import sys;{code}");
-                    _pendingPathAdditions.Clear();
+                        _pendingEndOfPathAdditions.Clear();
+                    }
+
+                    // Update our path cache
+                    pathCache = PythonEngine.Eval("sys.path", null, locals.Handle).As<List<string>>();
+                    locals.Dispose();
```

Comment:
Would be sweet if we can directly avoid the need of the cache 👍

Reply:
We don't need it per say, but I noticed in testing that we were adding the same directories more than once. Seemed like a lightweight solution to keep us from  duplicating dirs on our PythonPath on accident.

Reply:
Could always get the list<strings> on the fly and remove any duplicates before appending/inserting 😁